backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/S3Service.java

```diff
@@ -33,46 +46,159 @@
-                .withPathStyleAccessEnabled(true)
-                .withClientConfiguration(clientConfiguration)
-                .withCredentials(new AWSStaticCredentialsProvider(awsCreds))
-                .build();
+        LOGGER.debug("minio client");
+        initializeMinioClient(cloudAccessKey, cloudSecretKey, awsRegion, minioEndpoint);
       }
+    } else if (ModelDBUtils.isEnvSet(ModelDBConstants.AWS_ROLE_ARN)
+        && ModelDBUtils.isEnvSet(ModelDBConstants.AWS_WEB_IDENTITY_TOKEN_FILE)) {
+      LOGGER.debug("temporary token based s3 client");
+      initializeS3ClientWithTemporaryCredentials(awsRegion);
     } else {
+      LOGGER.debug("environment credentials based s3 client");
       // reads credential from OS Environment
       s3Client = AmazonS3ClientBuilder.standard().withRegion(awsRegion).build();
     }
+  }
 
-    this.bucketName = cloudBucketName;
+  private void initializeMinioClient(
+      String cloudAccessKey, String cloudSecretKey, Regions awsRegion, String minioEndpoint) {
+    AWSCredentials awsCreds = new BasicAWSCredentials(cloudAccessKey, cloudSecretKey);
+    ClientConfiguration clientConfiguration = new ClientConfiguration();
+    clientConfiguration.setSignerOverride("AWSS3V4SignerType");
+
+    this.s3Client =
+        AmazonS3ClientBuilder.standard()
+            .withEndpointConfiguration(
+                new AwsClientBuilder.EndpointConfiguration(minioEndpoint, awsRegion.getName()))
+            .withPathStyleAccessEnabled(true)
+            .withClientConfiguration(clientConfiguration)
+            .withCredentials(new AWSStaticCredentialsProvider(awsCreds))
+            .build();
+  }
+
+  private void initializeS3ClientWithAccessKey(
+      String cloudAccessKey, String cloudSecretKey, Regions awsRegion) {
+    BasicAWSCredentials awsCreds = new BasicAWSCredentials(cloudAccessKey, cloudSecretKey);
+    this.s3Client =
+        AmazonS3ClientBuilder.standard()
+            .withRegion(awsRegion)
+            .withCredentials(new AWSStaticCredentialsProvider(awsCreds))
+            .build();
+  }
+
+  private void initializeS3ClientWithTemporaryCredentials(Regions clientRegion) {
+
+    try {
+      FetchTokenAndSchedule(clientRegion.toString());
+
+      Map<String, String> credentialMap = new HashMap<String, String>();
+      populateCredentialMap(
+          credentialMap,
+          ModelDBUtils.appendOptionalTelepresencePath(
+              System.getenv(ModelDBConstants.AWS_WEB_IDENTITY_TOKEN_FILE)));
+
+      // Create a BasicSessionCredentials object that contains the credentials you just retrieved.
+      BasicSessionCredentials awsCredentials =
+          new BasicSessionCredentials(
+              credentialMap.get(ModelDBConstants.CLOUD_ACCESS_KEY),
+              credentialMap.get(ModelDBConstants.CLOUD_SECRET_KEY),
+              credentialMap.get(ModelDBConstants.AWS_WEB_IDENTITY_TOKEN));
```

Comment:
This looks weird to me. The flow is:

1. read the content of the file in a buffer (no need for parsing)
2. call STS method `AssumeRoleWithWebIdentity` (`AssumeRole` will not work)
3. use the resulting credentials

It seems like you're doing:

1. Read the file and parse it
2. Read the credentials from the file.

Reply:
this is fixed